lib/std/once.zig

```diff
@@ -7,65 +7,119 @@ const std = @import("std.zig");
 const builtin = std.builtin;
 const testing = std.testing;
 
-pub fn once(comptime f: fn () void) Once(f) {
+pub fn once(comptime f: anytype) Once(f) {
     return Once(f){};
 }
 
 /// An object that executes the function `f` just once.
-pub fn Once(comptime f: fn () void) type {
+pub fn Once(comptime f: anytype) type {
+    const T = @TypeOf(f);
+    const ty_info = @typeInfo(T);
+    if (ty_info != .Fn)
+        @compileError("expected function type, found " ++ @typeName(T));
+
+    const fn_info = ty_info.Fn;
+    if (fn_info.is_generic)
+        @compileError("cannot instantiate Once with a generic function");
+    if (fn_info.is_var_args)
+        @compileError("cannot instantiate Once with a variadic function");
+
+    // The return type must be void or !void.
+    const Return = fn_info.return_type.?;
+    if (!switch (@typeInfo(Return)) {
+        .Void => true,
+        .ErrorUnion => |info| info.payload == void,
+        else => false,
+    })
```

Comment:
style nitpick: maybe add { } around the if-body to help distinguish that its apart apart of the `if` given the condition is also indented.